pest/src/parser_state.rs

```diff
@@ -858,23 +858,67 @@ impl<'i, R: RuleType> ParserState<'i, R> {
     /// assert_eq!(result.unwrap().position().pos(), 2);
     /// ```
     #[inline]
-    pub fn stack_pop(self: Box<Self>) -> ParseResult<Box<Self>> {
-        let result_state = {
-            let string = self.stack
-                .peek()
-                .expect("pop was called on empty stack")
-                .as_str();
-
-            self.match_string(string)
-        };
+    pub fn stack_pop(mut self: Box<Self>) -> ParseResult<Box<Self>> {
+        let string = self.stack
+            .pop()
+            .expect("pop was called on empty stack")
+            .as_str();
+        self.match_string(string)
+    }
 
-        match result_state {
-            Ok(mut state) => {
-                state.stack.pop();
-                Ok(state)
-            }
-            Err(state) => Err(state)
+    /// Matches the full state of the stack.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use pest;
+    /// # #[allow(non_camel_case_types)]
+    /// # #[derive(Clone, Copy, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
+    /// enum Rule {}
+    ///
+    /// let input = "aaaa";
+    /// let mut state: Box<pest::ParserState<Rule>> = pest::ParserState::new(input);
+    /// let mut result = state.stack_push( |state| state.match_string("a")).and_then(|state| {
+    ///     state.stack_push( |state| state.match_string("a"))
+    /// }).and_then(|state| state.stack_match_peek());
+    /// assert!(result.is_ok());
+    /// assert_eq!(result.unwrap().position().pos(), 4);
+    /// ```
+    #[inline]
+    pub fn stack_match_peek(self: Box<Self>) -> ParseResult<Box<Self>> {
+        let spans = self.stack.get_state();
+        let string = spans.iter()
```

Comment:
Since you're iterating anyway, why not use [`Iterator::all`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.all)?

Reply:
I guess I could run `self.match_string()` multiple times. I'm not sure if that's better or worse than producing a string and matching against that.

Reply:
Joining the strings call O(N) memcpy and O(1) memcmp. Iterating should be cheaper because you run memcmp O(N) times and that's it.

Reply:
So the more I think about it, the more I like my initial solution. Mostly because repeatedly running `self.match_string()` involves wrapping and unwrapping subsequent `ParseResult`s. Also `match_string()` is consuming so I'm not entirely sure how to get `Iterator::fold` to work with it since the `self` is borrowed while iterating `self.stack`

Reply:
I agree with you. However, since this is implemented in `parser_state`, why not use `Position::match_string`? Then you could simply fold on a `bool`.

Reply:
Generally, I think that the solution is fine, but allocating a lot of small `Vec`s on the heap for every `peek_all` will slow down heap allocation after a while.